### ISSUE TITLE
Performance issue with memset() and memcpy() for large arrays

### ISSUE THREAD

**Issue by invd** [ultimate-pa/ultimate#308]
During recent analysis of Ultimate runtime performance against parts of the Trezor codebase, slow to very slow (several hours to several dozen hours) verification of comparably simple programs has been observed for reasonable and optimized toolchain/setting parameters.

In discussion with @danieldietsch and @jhoenicke, it has been suggested that this is due to Ultimates slow initialization / handling of data structures with a size above a few bytes, for example large buffers and strings.
During research of issue #303, the following example was used:

[memset_example_2.c](https://github.com/ultimate-pa/ultimate/blob/dev/trunk/examples/CToBoogieTranslation/trezor/memset_example_2.c)
```
#include <stdint.h>
// memset definition to satisfy gcc --pedantic
#include <string.h>

typedef unsigned char uint8_t;

static uint8_t _oledbuffer[8];

int main()
{
  memset(_oledbuffer, 0, sizeof(_oledbuffer));
  return 0;
}
```
Ultimate is able to run it's verification over this program in under a minute, with the following results and corresponding number of result occurences:
```
      2   - PositiveResult [Line: -1]: pointer dereference always succeeds
     16   - PositiveResult [Line: 1]: pointer dereference always succeeds
      1   - PositiveResult [Line: 9]: all allocated memory was freed
```

If the _oledbuffer array is initialized and then zeroed in the actual code for it's original purpose, which is holding the central framebuffer datastructure for a 128*64 1-bit OLED, then it needs to hold **1024** individual uint8_t elements (=1024 bytes) instead of just **8** elements used in the example above:
[memset_example_1.c](https://github.com/ultimate-pa/ultimate/blob/dev/trunk/examples/CToBoogieTranslation/trezor/memset_example_1.c)
```
#include <stdint.h>
// memset definition to satisfy gcc --pedantic
#include <string.h>

typedef unsigned char uint8_t;

static uint8_t _oledbuffer[1024];

int main()
{
  memset(_oledbuffer, 0, sizeof(_oledbuffer));
  return 0;
}
```
Ultimate now takes about **80 minutes** (!) on a modern system [^1] to run it's verification after this change, as it checks about 2000 potential error locations:
```
      2   - PositiveResult [Line: -1]: pointer dereference always succeeds
   2048   - PositiveResult [Line: 1]: pointer dereference always succeeds
      1   - PositiveResult [Line: 7]: all allocated memory was freed
```
It can be seen that the number of **Line: 1** results (which are misleading, see #303) directly corresponds to twice the number of elements in the _oledbuffer. 
This assumption is further tested by replacing the relevant line with ```static uint8_t _oledbuffer[64];```, which generates matching results in under a minute:
```
      2   - PositiveResult [Line: -1]: pointer dereference always succeeds
    128   - PositiveResult [Line: 1]: pointer dereference always succeeds
      1   - PositiveResult [Line: 9]: all allocated memory was freed
```

Preliminary tests show a runtime in O(n²) with n=number of elements of the array for program verification (n=8,n=64,n=256,n=380,n=1024 were tested).

It is important to note that with a measured runtime of about **76 minutes**, performance does not improve notably when just one element is processed by memset as compared to all 1024 elements:
[memset_example_3.c](https://github.com/ultimate-pa/ultimate/blob/dev/trunk/examples/CToBoogieTranslation/trezor/memset_example_3.c)
```
#include <stdint.h>
// memset definition to satisfy gcc --pedantic
#include <string.h>

typedef unsigned char uint8_t;

static uint8_t _oledbuffer[1024];

int main()
{
  memset(_oledbuffer, 0, 1);
  return 0;
}
```
This run also shows the same large number of results as the second program example.

However, when performing a similar access with a regular array operation against the full-sized array, Ultimate **does not** show comparable performance issues and finishes verification in **under a minute**:
[memset_example_5.c](https://github.com/ultimate-pa/ultimate/blob/dev/trunk/examples/CToBoogieTranslation/trezor/memset_example_5.c)
```
#include <stdint.h>

typedef unsigned char uint8_t;

static uint8_t _oledbuffer[1024];

int main()
{
  _oledbuffer[0] = 0;
  return 0;
}
```
Similarly, assigning all elements the value of zero in a for() loop is also verifiable in under a minute:
[memset_example_7.c](https://github.com/ultimate-pa/ultimate/blob/dev/trunk/examples/CToBoogieTranslation/trezor/memset_example_7.c)
```
#include <stdint.h>

typedef unsigned char uint8_t;

static uint8_t _oledbuffer[1024];

int main()
{
  for(int i = 0; i < sizeof(_oledbuffer); i++)
  {
    _oledbuffer[i] = 0;
  }
  return 0;
}
```

Preliminary results indicate that the use of memcpy() has similar performance impact to memset, as the following program takes in the order of **80 minutes** to verify:
[memcpy_example_1.c](https://github.com/ultimate-pa/ultimate/blob/dev/trunk/examples/CToBoogieTranslation/trezor/memcpy_example_1.c)
```
#include <stdint.h>
// memcpy definition to satisfy gcc --pedantic
#include <string.h>

typedef unsigned char uint8_t;

static uint8_t _oledbuffer[1024];

int main()
{
  memcpy(_oledbuffer, "42", 2);
  return 0;
}
```

During further discussion with @danieldietsch and @jhoenicke, we have confirmed that the use of memset and memcpy will produce boogie intermediary code that requires many costly error checks due to Ultimate's internal memory handling where "manual" array value assignments or loops do not.

Toolchain: 
* releaseScripts/default/UAutomizer-linux/config/AutomizerMemDerefMemtrack.xml

Settings tested:
* trunk/examples/settings/default/automizer/svcomp-DerefFreeMemtrack-32bit-Automizer_Default.epf

Versions affected: including Ultimate 0.1.23-846d8eb (newest).

[^1]: Debian Stretch KVM virtual machine (7 threads, 40GB RAM) running on a Linux Hypervisor w/ Intel i7-6700 CPU @ 3.40GHz and 4x16GB DDR4 2133 MHz RAM.

**Comment by danieldietsch** [comment]
One part of this issue should be addressed by 277cfe1: We now use a unchecked write function during initialization of global arrays.

**Comment by invd** [comment]
The behavior of Ultimate appears to have significantly improved with the mentioned fixes and other performance improvements.

The runtime of the tested example programs is now under 1 minute on the previously described VM: 

```
time ./Ultimate -tc ~/ultimate/releaseScripts/default/UAutomizer-linux/config/AutomizerMemDerefMemtrack.xml -s ~/ultimate/trunk/examples/settings/default/automizer/svcomp-DerefFreeMemtrack-32bit-Automizer_Default.epf -i ~/ultimate/trunk/examples/CToBoogieTranslation/trezor/memcpy_example_1.c
real    0m11.628s
user    0m24.840s
sys     0m1.256s
```

```
time ./Ultimate -tc ~/ultimate/releaseScripts/default/UAutomizer-linux/config/AutomizerMemDerefMemtrack.xml -s ~/ultimate/trunk/examples/settings/default/automizer/svcomp-DerefFreeMemtrack-32bit-Automizer_Default.epf -i ~/ultimate/trunk/examples/CToBoogieTranslation/trezor/memset_example_1.c
real    0m8.123s
user    0m19.312s
sys     0m0.880s
```

```
time ./Ultimate -tc ~/ultimate/releaseScripts/default/UAutomizer-linux/config/AutomizerMemDerefMemtrack.xml -s ~/ultimate/trunk/examples/settings/default/automizer/svcomp-DerefFreeMemtrack-32bit-Automizer_Default.epf -i ~/ultimate/trunk/examples/CToBoogieTranslation/trezor/memset_example_2.c
real    0m2.422s
user    0m5.156s
sys     0m0.172s
```

Memset examples 3 to 10 behave similarly well. The out of bounds access in memset_example_10.c is quickly detected. Therefore, the issue appears to be solved.
